### PR TITLE
[LoRA] feat: `save_lora_adapter()`

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Complementing https://github.com/huggingface/diffusers/pull/9712, this PR adds a `save_lora_adapter()` for the models that support LoRA loading. It also adds tests to ensure things don't break. 

Additionally, this PR:

* Deprecates the `load_attn_procs()` method when it tries to load a LoRA state dict (and adds tests for it). 
* Replaces the `load_attn_procs()` method with `load_lora_adapter()` in `src/diffusers/loaders/lora_pipeline.py`. I have run the integration tests for the SD and SDXL LoRAs (as those are impacted by this change), and the tests passed. 
* Some minor code improvements on the loading side (comments in line). 

Additionally, just to be sure, I have run the integration tests under `tests/lora` to ensure they pass.

### CODE REVIEW

File: check_flux.py
Comment (sayakpaul) [inline check_flux.py:1]: Ignore.

File: tests/models/test_modeling_common.py
Comment (sayakpaul) [inline tests/models/test_modeling_common.py:880]: Unrelated but my hands were itching.

File: tests/models/test_modeling_common.py
Comment (sayakpaul) [inline tests/models/test_modeling_common.py:965]: @BenjaminBossan do you have an idea as why these two state dicts have different lengths? IMO, they should have the exact same length? Or am I being mistaken here?

The keys that are missing in the retrieved state dict have the `"single_transformer_blocks"` prefix. 

I am on the `peft` main (`0.13.3.dev0`).

Comment (BenjaminBossan) [reply tests/models/test_modeling_common.py:965]: I debugged and the cause is this line:

https://github.com/huggingface/diffusers/pull/9862/files#diff-fdc2e14a8c1091d917cf31ca155be9baa8095d704bd2a2f58970ec31641c0caeR198

Since `prefix` is defined as `"transformer"` but the keys are starting with `"single_transformer_blocks"`, they don't match. In fact, even the other keys should not match: Even though they start with `"transformer"`, they don't start with `"transformer."` (note the dot), which _I think_ is the intended condition, since the next line replaces `f"{prefix}."`.

I think what happens here is that since this is a standalone module, not a transformer submodule of a larger model, it does not have a `"transformer."`  prefix. Therefore, the keys should not be filtered like that. Probably you know best what the correct solution would be here, but one suggestion is to make the `prefix` argument optional and only filter if it is not `None`.

Also, this problem probably rarely arise in real use cases, as users will typically not load the transformer as a standalone module.

Comment (sayakpaul) [reply tests/models/test_modeling_common.py]: Thank you very much!

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:201]: Could be a set comprehension for a little bit extra performance, but probably not much of a difference overall.

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:223]: Hmm, AFAICT the two conditions `k.startswith(f"{prefix}.") and k.split(".")[0] == prefix` are redundant, right? In the previous code, we had `k.startswith(prefix)`, which is a bit different, but was also unnecessary.

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:230]: This is the same code as before, just split into 2 lines, right?

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:296]:
```suggestion
        weight_name: Optional[str] = None,
```

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:310]: TIL about `.to(None)`.

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:314]: Does it make sense to return here instead of raising an error?

File: src/diffusers/loaders/peft.py
Comment (BenjaminBossan) [inline src/diffusers/loaders/peft.py:333]: I wonder if it would make sense to (optionally) save the peft config too or not.

File: tests/lora/utils.py
Comment (BenjaminBossan) [inline tests/lora/utils.py:1795]: Remove?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for adding a method to save the LoRA adapter. Overall, this looks good, I have a few comments but no blockers.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [reply src/diffusers/loaders/peft.py:333]: We could but we don't really support the loading. So, maybe something to consider for the future? I have made a note about it.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [reply src/diffusers/loaders/peft.py:314]: Changed to a `ValueError`

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [reply src/diffusers/loaders/peft.py:201]: Yeah.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [reply src/diffusers/loaders/peft.py:223]: Could just keep it to "k.startswith(f"{prefix}.")"?

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [reply src/diffusers/loaders/peft.py:230]: Yeah.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [inline src/diffusers/loaders/peft.py:201]: Just a better and more robust way to filter out the state dict.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [inline src/diffusers/loaders/peft.py:209]: Catching this error earlier than previous.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [inline src/diffusers/loaders/peft.py:222]: Removing redundant conditions.

File: src/diffusers/loaders/peft.py
Comment (sayakpaul) [inline src/diffusers/loaders/peft.py:234]: Breaking the conditionals to be more explicit.

File: tests/models/unets/test_models_unet_2d_condition.py
Comment (sayakpaul) [inline tests/models/unets/test_models_unet_2d_condition.py:1081]: Unneeded to test here now.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for this PR, LGTM. I assume the docstring will be completed before merging.

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽

**Comment by sayakpaul** [comment]
@yiyixuxu could you give this a review?

**Comment by bghira** [comment]
why isn't this merged yet?

**Comment by sayakpaul** [comment]
Pending reviews from Yiyi. It will take a bit as she is off for some days.

**Comment by yiyixuxu** [comment]
thanks!

**Pull request merged**
